src/mesh/mesh_adaptation.cpp

```diff
@@ -0,0 +1,117 @@
+#include "mesh_adaptation.h"
+
+namespace PHiLiP {
+
+template <int dim, typename real, typename MeshType>
+MeshAdaptation<dim,real,MeshType>::MeshAdaptation(double critical_res_input, int total_ref_cycle, double refine_frac, double coarsen_frac)
+    : critical_residual(critical_res_input)
+    , total_refinement_cycles(total_ref_cycle)
+    , refinement_fraction(refine_frac)
+    , coarsening_fraction(coarsen_frac)
+    , current_refinement_cycle(0)
+    , pcout(std::cout, dealii::Utilities::MPI::this_mpi_process(MPI_COMM_WORLD)==0)
+    {}
+
+template <int dim, typename real, typename MeshType>
+void MeshAdaptation<dim,real,MeshType>::adapt_mesh(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    if (!refine_mesh)
+    {
+        return;
+    }
+
+    if (total_refinement_cycles == current_refinement_cycle)
```

Comment:
This check is now performed in ODESolver.

Since adapt_mesh() adapts meshes just once, some sort of loop must exist outside wherever it's called for multiple refinements. Members `total_refinement_cycles` and `current_refinement_cycles` are still members of MeshAdaptation, but they are not used by the member functions and simply exist for book-keeping purposes (to make it easy for someone calling the mesh_adaptation object in a loop to set conditions to exit the loop).